daemon/runtime_unix.go

```diff
@@ -110,6 +113,19 @@ func (daemon *Daemon) initRuntimes(runtimes map[string]types.Runtime) (err error
 				}
 			}
 			rt.ShimConfig = defaultV2ShimConfig(daemon.configStore, daemon.rewriteRuntimePath(name, rt.Path, rt.Args))
+			var featuresStderr bytes.Buffer
+			featuresCmd := exec.Command(rt.Path, append(rt.Args, "features")...)
+			featuresCmd.Stderr = &featuresStderr
+			if featuresB, err := featuresCmd.Output(); err != nil {
+				logrus.WithError(err).Warnf("Failed to run %v: %q", featuresCmd.Args, featuresStderr.String())
+			} else {
+				var features features.Features
+				if jsonErr := json.Unmarshal(featuresB, &features); jsonErr != nil {
+					logrus.WithError(err).Warnf("Failed to unmarshal the output of %v as a JSON", featuresCmd.Args)
+				} else {
+					rt.Features = &features
+				}
+			}
```

Comment:
We will also need a way to populate the features of "runtimeType runtimes," a.k.a. containerd shims. Otherwise rro mounts will not be possible using runtimes which are not drop-in replacements for runC, e.g. kata-containers.

How feasible would it be to probe for support dynamically, by trying to start a pseudo-container from a spec containing an `rro` mount with each runtime?

Reply:
> kata

Probably “rro” doesn’t make much sense for VM runtimes, as “ro” is already recursive for virtiofs 

> pseudo 

Sounds overengineerring.
containerd should just provide the proper API for inspecting runtime features.

Reply:
> We will also need a way to populate the features of "runtimeType runtimes," a.k.a. containerd shims.

- PR: https://github.com/containerd/containerd/pull/8509